lib/eex/lib/eex/compiler.ex

```diff
@@ -478,4 +503,60 @@ defmodule EEx.Compiler do
     # length(~c"<%") == 2
     column + 2 + length(mark)
   end
+
+  defp code_snippet_with_previous_lines(state, line_end) do
+    state.tokens
+    |> Enum.take(3)
+    |> Enum.reverse()
+    |> case do
+      [{:expr, _, _, %{column: _c, line: line_start}} | _rest] ->
+        code_snippet(state.source, {line_start, line_end})
+
+      [] ->
+        code_snippet(state.source, {1, line_end})
+    end
+  end
+
+  defp code_snippet(source, {line_start, line_end}, opts \\ []) do
+    arrow_padding = Keyword.get(opts, :arrow_padding, 0)
+    {offset_start, offset_end} = source_offset(source, line_start, line_end)
+
+    {snippet, _acc} =
+      source
+      |> binary_part(offset_start, offset_end - 1)
+      |> String.split(["\r\n", "\n"])
+      |> Enum.map_reduce(line_start, fn
+        expr, line_number when line_number == line_end ->
+          spaces = count_until_expr(expr, 0)
+          arrow = String.pad_leading("^", spaces + 4 + arrow_padding)
+          expr = "#{line_number} | #{expr}\n  | #{arrow}"
+          {expr, line_number + 1}
+
+        expr, line_number ->
+          {"#{line_number} | #{expr}", line_number + 1}
+      end)
+
+    "  |\n#{Enum.map_join(snippet, "\n", & &1)}"
+  end
+
+  defp count_until_expr(<<char, rest::binary>>, count) when char in [?\s, ?\t],
+    do: count_until_expr(rest, count + 1)
+
+  defp count_until_expr(<<"<", _rest::binary>>, count), do: count
```

Comment:
We can't do this because we may have false positives, like this:

```
text <%= :example %> <%= if true do %>
```

or the opposite:

```
text <%= if true do %> <%= :example %> 
```

Is there a reason why you need to count until expression? Isn't the expression column enough/correct?